### ISSUE TITLE
How can I identify level1 nvswitch and level2 nvswitch in NCCL

### ISSUE THREAD

**Issue by Ryan201802** [NVIDIA/nccl#1286]
The different NVSwitches are not visible to NCCL. Not for NVLink communication, not for NVLink SHARP. Traffic is spread on all switches transparently.

_Originally posted by @sjeaugey in https://github.com/NVIDIA/nccl/issues/1006#issuecomment-1727221412_
            
I am confused how to use two levels nvswitch to do allreduce in NCCL.

**Comment by Ryan201802** [comment]
Thank for your help. But @AddyLaddy , how can I understand "inter-node NVLink SHARP since 2.18" mentioned at https://github.com/NVIDIA/nccl/issues/895#issuecomment-1606863525
how does the inter-node connect?

**Comment by Ryan201802** [comment]
Additionally, I find the framework in the GH200 white paper. There is two levels nvswitch.  
![image](https://github.com/NVIDIA/nccl/assets/36202248/20c787cb-8f2a-47c8-a93f-2725a00ba6ee)
In this framework, can the second level do NVLink SHARP?

**Comment by AddyLaddy** [comment]
We have developed so called Multi-Node NVLink systems (MNNVL) and the first publicly available system will be called GB200 (sometimes referred to as NVL72). 
[GB200 NVL72](https://www.nvidia.com/en-us/data-center/gb200-nvl72/)
It will have 72 Blackwell generation GPUs connected to a single NVLink domain.

**Comment by AddyLaddy** [comment]
The full features of NVLink will be available to all GPUs in such a system, including NVLink SHARP.
NCCL already supports NVLink SHARP on 8x H100 systems, so it's just the same but with a larger NVLink domain. 
It's all opaque to NCCL, it just sees 72 GPUs all accessible via NVLink.

**Comment by Ryan201802** [comment]
Wow, the NVL72 is great product. Does it support multi-level nvswitch?

**Comment by Ryan201802** [comment]
Thanks, I understand. 
Now only have a question about the GH200 Framework. How does the second level nvswitch work? or how can I use their Sharps unit?

**Comment by AddyLaddy** [comment]
I don't believe we've announced the NVLink topology of GB200 yet.
But NVLink SHARP works in both single and two level NVSwitch networks.
Again, it's all opaque to NCCL. It just works like any other NVLink connected machine that is NVLink SHARP capable.

**Comment by hennry205** [comment]
Hi @AddyLaddy, I have a question. How does multimem.ld_reduce command in kernel trigger nvswitch to do Reduce function? Because I don't see any direct control of nvswith in the kernel, Thanks.

**Comment by Ryan201802** [comment]
Not GB200, it's GH200 mentioned in the previous discussion

**Comment by Ryan201802** [comment]
Yes @AddyLaddy , but the topology I mentioned is **GH200**, not GB200.

**Comment by AddyLaddy** [comment]
There are no publicly released Nvidia products that use multi level NVSwitches currently. The first publicly available product will be GB200.
NVLink SHARP works with both single level and multi-level NVLink fabrics.

**Comment by AddyLaddy** [comment]
> Hi @AddyLaddy, I have a question. How does multimem.ld_reduce command in kernel trigger nvswitch to do Reduce function? Because I don't see any direct control of nvswith in the kernel, Thanks.

The NVLink SHARP implementation is based in HW and is configured by mapping in special "Multicast" enabled buffer addresses to the GPU virtual address space. See [CUDA Multicast](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MULTICAST.html)

**Comment by shanleo1986** [comment]
@AddyLaddy Can you help to explain more about this comments?https://github.com/NVIDIA/nccl/issues/1159#issuecomment-1908823285
I have some questions about MNNVL.
(1) NCCL is designed to search topo locally, how one host can read/write the GPU buffer belongs to the other host? How can they know each other?
(2) The buffer alloced by cuMemCreate can be used by cuMemImportFromShareableHandle on another host?

Thank you!

**Comment by kkndyu** [comment]
> Hi [@AddyLaddy](https://github.com/AddyLaddy), I have a question. How does multimem.ld_reduce command in kernel trigger nvswitch to do Reduce function? Because I don't see any direct control of nvswith in the kernel, Thanks.

Here I guess:

There must be some NVlink protocol level support for SHARP, the param mc_handle of multimem.ld_reduce records infos of rank, nranks, addrs, and lens etc, execution of this instruction will trigger the filling of these items to related field of NVlink packet head, when switch get these kind of collective packets, it will analyse the fileds and load the buffers and reduce ...
